metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1022,7 +1020,13 @@ public void handleLoadSnapshot(SnapshotReader<ApiMessageAndVersion> reader) {
                         Batch<ApiMessageAndVersion> batch = reader.next();
                         long offset = batch.lastOffset();
                         List<ApiMessageAndVersion> messages = batch.records();
-
+                        if (reader.snapshotId().equals(Snapshots.BOOTSTRAP_SNAPSHOT_ID)) {
+                            // For bootstrap snapshots, extract feature levels from all data records
+                            if (!messages.isEmpty()) {
+                                bootstrapMetadata = BootstrapMetadata.fromRecords(messages, "bootstrap");
+                                return;
+                            }
+                        }
```

Comment:
Let's add an in-line comment describing what we're doing here for KIP-1170, since it is not super obvious to the reader.

```suggestion
// KIP-1170: The 0-0.checkpoint can contain metadata records. If it does, they should be considered the bootstrap metadata for the cluster.
                        if (reader.snapshotId().equals(Snapshots.BOOTSTRAP_SNAPSHOT_ID) && !messages.isEmpty()) {
                                bootstrapMetadata = BootstrapMetadata.fromRecords(messages, "bootstrap");
                                return;
                        }
```